sandy/core/cov.py

```diff
@@ -393,20 +534,216 @@ def decompose(self):
+        threshold : `int`, optional
+            threshold to avoid numerical fluctuations. The default is None.
+
+        Returns
+        -------
+        `CategoryCov`
+            GlS sensitivity for a given Vy and S.
+
+        Example
+        -------
+        >>> S = np.array([[1, 2], [3, 4]])
+        >>> var = sandy.CategoryCov.from_var([1, 1])
+        >>> var._gls_sensitivity(S, np.array([1, 1]))
+                      0	          1
+        0	2.57143e-01	3.14286e-01
+        1	3.14286e-01	8.28571e-01
+
+        >>> S = pd.DataFrame([[1, 2], [3, 4]], columns =[1, 2],index=[3, 4])
+        >>> var = sandy.CategoryCov.from_var([1, 1])
+        >>> var._gls_sensitivity(S,pd.Series([1, 1], index=[1, 2]))
+                      1	          2
+        1	2.57143e-01	3.14286e-01
+        2	3.14286e-01	8.28571e-01
+        """
+        columns = pd.DataFrame(S).columns
+        S_ = pd.DataFrame(S).values
+        Vy_ = pd.DataFrame(np.diag(pd.Series(Vy))).values
+        # GLS_sensitivity:
         Vx = self.data.values
-        V = S.T.dot(Vx.dot(S)) + Vy
-        C = Vx - Vx.dot(S.dot(np.linalg.inv(V).dot(S.T.dot(Vx))))
-        return self.__class__(C)
-
-    def sandwich(self, Si, Sj=None):
-        C = self.data.values
-        Si_ = Si.reshape(self.size, -1)
-        if Sj is None:
-            Sj_ = Si_
-        else:
-            Sj_ = Sj.reshape(self.size, -1)
-        return Si_.T.dot(C.dot(Sj_))
+        M = S_.T.dot(Vx).dot(S_) + Vy_
+        M_inv = sandy.CategoryCov(M).invert()
+        sensitivity = Vx.dot(S_).dot(M_inv.data.values).dot(S_.T)
+        if threshold is not None:
+            sensitivity[sensitivity < threshold] = 0
+        return pd.DataFrame(sensitivity, index=columns, columns=columns)
+
+    def gls_update(self,  S, Vy, threshold=None):
+        """
+        Perform GlS update for a given variance and sensitivity.
+
+        Parameters
+        ----------
+        Vy : 1D iterable
+            Extra Covariance vector (MX1).
+        S : 2D iterable
+            Sensitivity square matrix (MXM).
+        delta : 1D iterable
+            Perturbed vector minus non perturbed vector.
+        threshold : `int`, optional
+            Thereshold to avoid numerical fluctuations. The default is None.
+
+        Returns
+        -------
+        `CategoryCov`
+            GLS method apply to a CategoryCov object for a given Vy and S.
+
+        Example
+        -------
+        >>> S = np.array([[1, 2], [3, 4]])
+        >>> var = sandy.CategoryCov.from_var([1, 1])
+        >>> var.gls_update(S, pd.Series([1, 1],index=[1, 2]))
+                     0            1
+        0  7.42857e-01 -3.14286e-01
+        1 -3.14286e-01  1.71429e-01
+        """
+        index, columns = self.data.index, self.data.columns
+        Vx = self.data.values
+        A = self._gls_sensitivity(S, Vy).values
+        V_new = Vx - A.dot(Vx)
+        V_new = pd.DataFrame(V_new, index=index, columns=columns)
+        if threshold is not None:

```

Comment:
this threshold is weird...I thought it was supposed to be passed to `_gls_sensitivity`.
Where do we need the threshold? on the variance or on the sensitivity?